pkg/client/api/auth.go

```diff
@@ -70,42 +75,63 @@ func (c *auth) Refresh(refreshToken string) (*api.AuthResponse, error) {
 }
 
 func (c *auth) DeviceCode(authKind, slugID string) (*oauth2.DeviceAuthResponse, error) {
-	result := &oauth2.DeviceAuthResponse{}
+	config := oauth2.Config{
+		Endpoint: oauth2.Endpoint{
+			DeviceAuthURL: c.deviceAuthURL(authKind, slugID),
+		},
+	}
 
-	return result, c.client.Post().
-		APIVersion("").
-		Resource(fmt.Sprintf("%s/providers/%s/%s/device/code", authResource, authKind, slugID)).
-		Do().
-		Object(result)
-}
+	resp, err := config.DeviceAuth(context.Background())
 
-func (c *auth) DeviceAccessToken(authKind, slugID, deviceCode string) (*api.AuthResponse, error) {
-	body := &api.TokenRequest{
-		GrantType:  api.GrantTypeDeviceCode,
-		DeviceCode: deviceCode,
+	// Strangely enough, the oauth2.DeviceAuth function does not parse the error body, so we have to it ourselves
+	oauthErr := &oauth2.RetrieveError{}
+	if err != nil && errors.As(err, &oauthErr) {
+		unmErr := json.Unmarshal(oauthErr.Body, &oauthErr)
+		if unmErr != nil {
+			return nil, &perseshttp.RequestError{Err: unmErr}
+		}
+		return nil, &perseshttp.RequestError{Err: oauthErr}
 	}
-	result := &api.AuthResponse{}
 
-	return result, c.client.Post().
-		APIVersion("").
-		Resource(fmt.Sprintf("%s/providers/%s/%s/token", authResource, authKind, slugID)).
-		Body(body).
-		Do().
-		Object(result)
+	return resp, nil
+}
+
+func (c *auth) DeviceAccessToken(authKind, slugID string, deviceAuthResp *oauth2.DeviceAuthResponse) (*api.AuthResponse, error) {
+	config := &oauth2.Config{
+		Endpoint: oauth2.Endpoint{
+			TokenURL: c.tokenURL(authKind, slugID),
+		},
+	}
+	token, err := config.DeviceAccessToken(context.Background(), deviceAuthResp)
+	if err != nil {
+		return nil, &perseshttp.RequestError{Err: err}
+	}
+	return &api.AuthResponse{
+		AccessToken:  token.AccessToken,
+		RefreshToken: token.RefreshToken,
+	}, nil
 }
 
 func (c *auth) ClientCredentialsToken(authKind, slugID, clientID, clientSecret string) (*api.AuthResponse, error) {
-	body := &api.TokenRequest{
-		GrantType:    api.GrantTypeClientCredentials,
+	config := &clientcredentials.Config{
```

Comment:
This method should not be used out of the percli context actually. 

If a user want to programmatically request perses (with the GO SDK for example), this should be something like that. 

For now this config is not integrated into the ``RestConfigClient`` struct but we might want to add it.
```go
  client, err := perseshttp.NewFromConfig(config.RestConfigClient)
	if err != nil {
		return nil, err
	}

  oauthConfig := &clientcredentials.Config{
		ClientID:     clientID,
		ClientSecret: clientSecret,
		TokenURL:     c.tokenURL(authKind, slugID),
		AuthStyle:    oauth2.AuthStyleInHeader,
	}
	client.Client = oauthConfig.Client(context.Background())
  return client, err
```

Reply:
that's where I'm struggling to understand your point.

the perses http client could be used (like you mention) with the usage of the GO SDK. Since this method is exposed I don't get why it is supposed to work only for the CLI.

If it is really the case, then we have an issue somewhere. the package `client/api` is supposed to be a standalone library

Reply:
It works for sure, but I'm telling it's designed for the CLI 

Because when we are in a golang code, that have a persistent context in memory we should use the golang oauth2 library to generate an http client that **automatically** manages the generating of the token when requesting. That's to comply with the specification of that lib that I'm doing these changes.